MONGOID-5222 Make mongoization of uncastable values consistent across types and document (#5319)
mongodb/mongoid: docs/reference/fields.txt

alias can :ref:`be removed <unalias-id>` if desired (such as to integrate
with systems that use the ``id`` field to store value different from ``_id``.

.. _uncastable-values:

Assigning Uncastable Values
---------------------------

In Mongoid 8, Mongoid has standardized the treatment of the assignment of
"uncastable" values. A value is considered "uncastable" when it cannot be

> coerced to the type of the assigning field. For example:

Reviewer: I would just say "to the type of the field"

Author: done

.. code::

  class User
    include Mongoid::Document

    field :name, type: Integer
  end

  User.new(name: [ "hello" ])

Assigning an array to a field of type Integer doesn't work since an array can't
be coerced to an Integer. To deal with this case, Mongoid has introduced the
``validate_attribute_types`` flag.

When the ``validate_attribute_types`` flag is turned on, the assignment of uncastable

> values to a field will cause an ``InvalidValue`` error to be raised.

Reviewer: ```suggestion
values to a field will cause a ``Mongoid::Errors::InvalidValue`` exception being raised.
```

   venue = Venue.new(location: point) # This uses the Point#mongoize instance method.
   venue = Venue.new(location: [ 12, 24 ]) # This uses the Point.mongoize class method.

.. note::

>   The ``mongoize`` method should raise an "InvalidValue" error on values that

Reviewer: ```suggestion
  The ``mongoize`` method should raise a ``Mongoid::Errors::InvalidValue`` on values that
```

>   are uncastable to your custom type. Mongoid will handle catching that error

Reviewer: Suggest "Mongoid will handle that error and write `nil`"  since "catch" is an operation in Ruby that is different from "rescue".

Author: done
